ledger/participant-integration-api/src/main/scala/platform/apiserver/error/ErrorInterceptor.scala

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.apiserver.error
+
+import com.daml.error.{BaseError, DamlContextualizedErrorLogger}
+import com.daml.error.definitions.LedgerApiErrors
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import io.grpc.ForwardingServerCall.SimpleForwardingServerCall
+import io.grpc.{
+  ForwardingServerCallListener,
+  Metadata,
+  ServerCall,
+  ServerCallHandler,
+  ServerInterceptor,
+  Status,
+  StatusException,
+  StatusRuntimeException,
+}
+
+final class ErrorInterceptor extends ServerInterceptor {
+
+  private val logger = ContextualizedLogger.get(getClass)
+  private val emptyLoggingContext = LoggingContext.newLoggingContext(identity)
+
+  override def interceptCall[ReqT, RespT](
+      call: ServerCall[ReqT, RespT],
+      headers: Metadata,
+      next: ServerCallHandler[ReqT, RespT],
+  ): ServerCall.Listener[ReqT] = {
+
+    val forwardingCall = new SimpleForwardingServerCall[ReqT, RespT](call) {
+
+      /** Here we are trying to detect status/trailers pairs that:
+        * - originated from the server implementation (i.e. Participant services as opposed to internal to gRPC implementation) AND
+        * - did not originate from self-service errors infrastructure.
+        * TODO error codes: Do we need to inspect/sanitize errors internal to gRPC implementation?
+        *
+        * We are handling unary endpoints that returned failed Futures,
+        * 2) streaming Akka endpoints that failed inside streaming.
+        * We are NOT handling here exceptions thrown outside of Futures or Akka streams. These are handled separately in
+        * [[com.daml.platform.apiserver.error.ErrorListener]].
+        * We are NOT handling here exceptions thrown inside Akka streaming. These are handled in
+        * [[com.daml.grpc.adapter.server.akka.ServerAdapter.toSink]]
+        *
+        * Details:
+        * The gRPC services that we generate via scalapb are using [[scalapb.grpc.Grpc.completeObserver]]
+        * when bridging from Future[T] and into io.grpc.stub.StreamObserver[T].
+        * [[scalapb.grpc.Grpc.completeObserver]] does the following:
+        * a) propagates instances of StatusException and StatusRuntimeException without changes,
+        * b) translates other throwables into a StatusException with Status.INTERNAL.
+        * We assume that we don't need to deal with a) but need to detect and deal with b).
+        * Knowing that Status.INTERNAL is used only by [[com.daml.error.ErrorCategory.SystemInternalAssumptionViolated]],
+        * which is marked a security sensitive, we have the following heuristic: check whether gRPC status is Status.INTERNAL
+        * and gRPC status description is not security sanitized.
+        * TODO error codes: Consider also checking whether metadata is security sanitized.
+        * TODO error codes: Consider intercepting the exceptions before they are translated in a Status/Metadata pair.
```

Comment:
Did you explore this possibility? Why is this a TODO?

Reply:
I'd like to get your opinion on that.
The one of doing that I see is writing our own scalapb plugin for unary services and/or replace scalapb's runtime library with our custom one (or shadow just the needed class).

For now I went we the approach of a custom server interceptor. If it is good enough than there might no point trying to do something else. In any case, the approach in this PR might serve as the first approximation to whatever we deem final.

Reply:
I suggest we leave it at that and let ourselves gather some experience with the current implementation. We don't really know if some unexpected errors will be leaking through the cracks, because well, those errors are unexpected.